rfcs/react-components/convergence/slot-children-render-function.md

```diff
@@ -0,0 +1,526 @@
+wrapping other slots inside of them.
+
+## Current issues
+
+[Slot children render function won't override existing children. #27089](https://github.com/microsoft/fluentui/issues/27089)
+
+Repro: https://codesandbox.io/s/elated-babbage-46ez1w?file=/example.tsx
+
+It's not possible to use children render function to override an entire slot, if `children` are already used. The above is a minimal repro of an issue that can be seen in Fluent in the `AccordionHeader` component. The component renders the `button` slot which contains its own children. JSX children will win over props children - which is how children render functions are rendered by `getSlots`
+
+https://github.com/microsoft/fluentui/blob/220b321042b4bcc126c86a91811a34b04934d49b/packages/react-components/react-accordion/src/components/AccordionHeader/renderAccordionHeader.tsx#L13-L25
+
+Requirements for children render function:
+
+- should be able to override the entire slot
+- When rendering the original children, should respect nested slots.
+
+## Problem statement
+
+The main problem revolves around coalescing those 3 instances of properties a slot might have! At the moment,
+our current slot declaration mechanism revolves around 2 separate methods used in 2 different moments:
+
+1. `resolveShorthand`
+2. `getSlots`
+
+`resolveShorthand` is invoked at `useState` hook, it will receive a shorthand and will convert it to either `undefined` or an object comprising
+the merging of the External properties and Internal ones, following given priorities (`props` wins over `defaultProps`).
+
+`getSlots` is invoked at `render` method and will go over every single slot declared on `state.components` and convert every resolved shorthand into a pair of
+element to be render and properties to be passed. In this `getSlots` method, we also ensure that if a given resolved shorthand `children`
+property is a function, than we'll invoke that function:
+
+```ts
+// getSlots
+
+if (typeof children === 'function') {
+  const render = children;
+  // in this case the pair of element to be render and properties to be passed are converted into
+  // a Fragment and {children} where `children` will be the return of invoking the `children` render method
+  return [
+    React.Fragment,
+    {
+      children: render(slot, propsWithoutAsAndChildren),
+    } as unknown as R[K],
+  ];
+}
+```
+
+### Problem 1
+
+The `slot` argument will represent the merging between `state.component.button` value
+(back into the `AccordionHeader` `button` slot example) and `props.button.as` value. `propsWithoutAsAndChildren`
+as the name suggests is all the properties provided by the `resolveShorthand` invocation without including `as` and also `children`
+(since `children` by itself is the render function).
+
+Since we lost the Internal properties on the merging between it and the External ones
+There's no way to provide such the default declared `children` to the render function!
```

Comment:
Yes, in this case we'll lose `defaultProps.children` which could be the argument to be passed down the line to the render function

Reply:
````suggestion
Since we lost the Internal properties on the merging between it and the External ones
There's no way to provide such the default declared `children` to the render function!

```js
function useComponent(props) {
  return {
    slot: resolveShorthand(props.slot, { children: 'foo' })
  }
}

const state = useComponent({ children: () => {} })
typeof state.slot.children === 'function' // true

// 💥 We lost `defaultProps.children` i.e. `children: "foo"`
```
````

Can you please add that code snippet to the RFC?